### ISSUE TITLE
Flatpak should spawn a terminal (if not already running in one) when launching Terminal=true applications

### ISSUE THREAD

**Issue by linuxhw** [flatpak/flatpak#4115]
## Linux distribution and version

Recent Fedora 33 with all latest updates.

## Description of the problem

A terminal window is not spawned when launching `Terminal=true` applications from the Alt+F2 dialogue in GNOME. The program is executed silently in a background process.

The issue was initially created for GNOME Software (https://gitlab.gnome.org/GNOME/gnome-software/-/issues/552#note_1032814), but seems that the fix needs to be in Flatpak instead.

## Steps to reproduce

1. Install `org.linux_hardware.hw-probe` by the following instructions: https://github.com/linuxhw/hw-probe#flatpak
2. Open Alt+F2 dialogue in GNOME
3. Run `flatpak run org.linux_hardware.hw-probe`
4. Nothing is shown on the screen

**Comment by refi64** [comment]
Isn't `Terminal=true` interpreted by the shell only? If you do the steps you stated with any locally installed CLI application, I believe a terminal still would not open, so there's nothing Flatpak is particularly doing wrong here...

**Comment by pwithnall** [comment]
> Isn't `Terminal=true` interpreted by the shell only?

Nope, `g_app_info_launch()` interprets it and prepends a terminal emulator to the command line if `Terminal=true`. See `prepend_terminal_to_vector()` in `gdesktopappinfo.c` in GLib. `g_app_info_launch()` is the common path for launching an application from a `.desktop` file.

> If you do the steps you stated with any locally installed CLI application, I believe a terminal still would not open, so there's nothing Flatpak is particularly doing wrong here...

The point is that this is being run with `flatpak run`, i.e. the flatpak code path for launching a flatpak. This situation is, I believe, analogous to the more common situation where someone’s installed a CLI flatpak via gnome-software and wants to launch it from there.

**Comment by smcv** [comment]
The `Terminal=true` annotation is in one of zero or more `.desktop` files, but by running `flatpak run` you're telling Flatpak to run the `command` from the app's manifest, which doesn't *necessarily* have anything to do with *any* of its `.desktop` files - although obviously in most cases, it would be silly for the `command` and the `Exec=` field in the main/only `.desktop` file to be different.

If you run `gio launch org.linux_hardware.hw-probe` (requires GLib 2.67.x), then *that* is launching a `.desktop` file, and will probably put it in a terminal.

Perhaps it would be close enough to assume that if the app has a `.desktop` file that matches the app ID, then `Terminal=true` in that `.desktop` file is actually applicable to the `command`?

What we don't want to happen is: `com.valvesoftware.Steam` installs a GUI app (the Steam client) and an accompanying CLI tool (`steamcmd`). I don't think it actually installs a `.desktop` file with `Terminal=true` for `steamcmd`, but imagine for a moment that it did. If you run `flatpak run com.valvesoftware.Steam`, then we certainly don't want to look at its two `.desktop` files, see that one of them has `Terminal=true`, and run the GUI Steam client (the default `command` from the manifest), in a terminal.

**Comment by linuxhw** [comment]
> Nope, `g_app_info_launch()` interprets it and prepends a terminal emulator to the command line if `Terminal=true`. See `prepend_terminal_to_vector()` in `gdesktopappinfo.c` in GLib. `g_app_info_launch()` is the common path for launching an application from a `.desktop` file.

Why this doesn't work if I run the app by `Launch` button in the Gnome-Software?

**Comment by linuxhw** [comment]
> If you run `gio launch org.linux_hardware.hw-probe` (requires GLib 2.67.x), then _that_ is launching a `.desktop` file, and will probably put it in a terminal.

How to attach this command to `Launch` button handler in Gnome-Software? Should I fix somehow `org.linux_hardware.hw-probe.yaml` or `hw-probe.sh`? See https://github.com/linuxhw/hw-probe/tree/master/flatpak

**Comment by pwithnall** [comment]
> Perhaps it would be close enough to assume that if the app has a `.desktop` file that matches the app ID, then `Terminal=true` in that `.desktop` file is actually applicable to the `command`?

Yes, or the manifest grows a key which points to the main `.desktop` file, or which is similar to `Terminal=true` itself.

Conversely, I don’t think we want the manifest or `.desktop` file to start explicitly listing a terminal emulator in the command, because that results in a second terminal emulator being run if you run `flatpak run …` from within a terminal, and forces the user to use a particular terminal emulator.

> Why this doesn't work if I run the app by `Launch` button in the Gnome-Software?

Because gnome-software calls `flatpak_installation_launch()`, not `g_app_info_launch()`.

> How to attach this command to Launch button handler in Gnome-Software? Should I fix somehow org.linux_hardware.hw-probe.yaml or hw-probe.sh? See https://github.com/linuxhw/hw-probe/tree/master/flatpak

I think we need some more input from flatpak developers before deciding what to do.
